File: packages/python/src/kingpepe_sdk/address.py
# Copyright (c) 2026 The KingPepe developers
# Distributed under the MIT software license, see the accompanying LICENSE file.
"""Address / Bech32 helpers.

These perform BASIC, self-contained format validation (Bech32/Bech32m structure
+ checksum + KingPepe HRP). They are NOT a substitute for consensus-critical
validation. For authoritative validation, ask a KingPepe Core node via the
``validateaddress`` RPC (see :meth:`KingPepeClient.validate_address`).
"""

from __future__ import annotations

from .errors import ValidationError
from .network import BITCOIN_HRPS, Network, network_for_hrp

_CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
_GENERATOR = (0x3B6A57B2, 0x26508E6D, 0x1EA119FA, 0x3D4233DD, 0x2A1462B3)


def _polymod(values: list[int]) -> int:
    chk = 1
    for v in values:
        top = chk >> 25
        chk = ((chk & 0x1FFFFFF) << 5) ^ v
        for i in range(5):
            if (top >> i) & 1:
                chk ^= _GENERATOR[i]
    return chk


def _hrp_expand(hrp: str) -> list[int]:
    return [ord(c) >> 5 for c in hrp] + [0] + [ord(c) & 31 for c in hrp]


def decode_bech32(value: str) -> tuple[str, str] | None:
    """Decode a Bech32/Bech32m string, verifying the checksum.

    Returns ``(hrp, encoding)`` where encoding is ``"bech32"`` or ``"bech32m"``,
    or ``None`` on failure.
    """
    if value != value.lower() and value != value.upper():
        return None
    s = value.lower()
    pos = s.rfind("1")
    if pos < 1 or pos + 7 > len(s) or len(s) > 90:
        return None
    hrp = s[:pos]
    data = []
    for c in s[pos + 1 :]:
        d = _CHARSET.find(c)
        if d == -1:
            return None
        data.append(d)
    chk = _polymod(_hrp_expand(hrp) + data)
    if chk == 1:
        return hrp, "bech32"
    if chk == 0x2BC830A3:
        return hrp, "bech32m"
    return None


def get_network_for_address(address: str) -> Network | None:
    """Return the KingPepe network a Bech32 address belongs to, or None."""
    decoded = decode_bech32(address)
    if not decoded:
        return None
    return network_for_hrp(decoded[0])


def is_bitcoin_address(address: str) -> bool:
    """True if the string is a Bitcoin Bech32 address (bc/tb/bcrt)."""
    decoded = decode_bech32(address)
    return bool(decoded and decoded[0] in BITCOIN_HRPS)


def is_valid_kingpepe_address_format(address: str, network: Network | None = None) -> bool:
    """Basic format validation for a KingPepe Bech32 address; rejects Bitcoin HRPs."""
    decoded = decode_bech32(address)
    if not decoded:
        return False
    addr_network = network_for_hrp(decoded[0])
    if addr_network is None:
        return False
    if network is not None and addr_network != network:
        return False
    return True


def assert_kingpepe_address(address: str, network: Network | None = None) -> None:
    """Assert a KingPepe address format, raising :class:`ValidationError` otherwise."""
    if is_bitcoin_address(address):
        raise ValidationError("Refusing a Bitcoin address in a KingPepe-facing operation.")
    if not is_valid_kingpepe_address_format(address, network):
        raise ValidationError(
            f"Not a valid KingPepe {network or ''} address (basic format check). "
            "Use the node's validateaddress RPC for authoritative validation."
        )

File: packages/javascript/src/address.ts
// Copyright (c) 2026 The KingPepe developers
// Distributed under the MIT software license, see the accompanying LICENSE file.

import { BITCOIN_HRPS, KingPepeNetwork, networkForHrp } from "./network.js";
import { ValidationError } from "./errors.js";

/**
 * Address / Bech32 helpers.
 *
 * These perform BASIC, self-contained format validation (Bech32/Bech32m
 * structure + checksum + KingPepe HRP). They are NOT a substitute for
 * consensus-critical validation. For authoritative validation, ask a KingPepe
 * Core node via the `validateaddress` RPC (see KingPepeClient.validateAddress).
 */

const CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const GENERATOR = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];

function polymod(values: number[]): number {
  let chk = 1;
  for (const v of values) {
    const top = chk >> 25;
    chk = ((chk & 0x1ffffff) << 5) ^ v;
    for (let i = 0; i < 5; i++) {
      if ((top >> i) & 1) chk ^= GENERATOR[i]!;
    }
  }
  return chk >>> 0;
}

function hrpExpand(hrp: string): number[] {
  const out: number[] = [];
  for (let i = 0; i < hrp.length; i++) out.push(hrp.charCodeAt(i) >> 5);
  out.push(0);
  for (let i = 0; i < hrp.length; i++) out.push(hrp.charCodeAt(i) & 31);
  return out;
}

type Bech32Decoded = { hrp: string; encoding: "bech32" | "bech32m" };

/** Decode a Bech32/Bech32m string, verifying the checksum. Returns null on failure. */
export function decodeBech32(input: string): Bech32Decoded | null {
  // Reject mixed case (BIP173).
  if (input !== input.toLowerCase() && input !== input.toUpperCase()) return null;
  const s = input.toLowerCase();
  const pos = s.lastIndexOf("1");
  if (pos < 1 || pos + 7 > s.length || s.length > 90) return null;
  const hrp = s.slice(0, pos);
  const dataPart = s.slice(pos + 1);
  const data: number[] = [];
  for (const c of dataPart) {
    const d = CHARSET.indexOf(c);
    if (d === -1) return null;
    data.push(d);
  }
  const chk = polymod([...hrpExpand(hrp), ...data]);
  if (chk === 1) return { hrp, encoding: "bech32" };
  if (chk === 0x2bc830a3) return { hrp, encoding: "bech32m" };
  return null;
}

/** Return the KingPepe network a Bech32 address belongs to, or undefined. */
export function getNetworkForAddress(address: string): KingPepeNetwork | undefined {
  const decoded = decodeBech32(address);
  if (!decoded) return undefined;
  return networkForHrp(decoded.hrp);
}

/** True if the string is a Bitcoin Bech32 address (bc/tb/bcrt). */
export function isBitcoinAddress(address: string): boolean {
  const decoded = decodeBech32(address);
  if (!decoded) return false;
  return (BITCOIN_HRPS as readonly string[]).includes(decoded.hrp);
}

/**
 * Basic format validation for a KingPepe Bech32 address. Optionally require a
 * specific network. Rejects Bitcoin HRPs. Returns true/false; does not throw.
 */
export function isValidKingPepeAddressFormat(address: string, network?: KingPepeNetwork): boolean {
  const decoded = decodeBech32(address);
  if (!decoded) return false;
  const addrNetwork = networkForHrp(decoded.hrp);
  if (!addrNetwork) return false; // not a KingPepe HRP (includes all Bitcoin HRPs)
  if (network && addrNetwork !== network) return false;
  return true;
}

/**
 * Assert a KingPepe address format, throwing {@link ValidationError} with a
 * specific reason. Use in payment/withdrawal flows to fail fast and clearly.
 */
export function assertKingPepeAddress(address: string, network?: KingPepeNetwork): void {
  if (isBitcoinAddress(address)) {
    throw new ValidationError("Refusing a Bitcoin address in a KingPepe-facing operation.");
  }
  if (!isValidKingPepeAddressFormat(address, network)) {
    throw new ValidationError(
      `Not a valid KingPepe ${network ?? ""} address (basic format check). ` +
        "Use the node's validateaddress RPC for authoritative validation.",
    );
  }
}
